snapcraft/internal/repo.py

```diff
@@ -115,28 +120,133 @@ def __init__(self, package_name):
         self.package_name = package_name
 
 
+class _AptCache:
+
+    def __init__(self, cache_dir, deb_arch, *,
+                 sources_list=None, use_geoip=False):
+        self._cache_dir = cache_dir
+        self._deb_arch = deb_arch
+        self._sources_list = sources_list
+        self._use_geoip = use_geoip
+
+    def _setup_apt(self, download_dir):
+        apt.apt_pkg.config.set("Dir::Cache::Archives", download_dir)
+
+        # Do not install recommends
+        apt.apt_pkg.config.set('Apt::Install-Recommends', 'False')
+
+        # Make sure we always use the system GPG configuration, even with
+        # apt.Cache(rootdir).
+        for key in 'Dir::Etc::Trusted', 'Dir::Etc::TrustedParts':
+            apt.apt_pkg.config.set(key, apt.apt_pkg.config.find_file(key))
+
+        # Clear up apt's Post-Invoke-Success as we are not running
+        # on the system.
+        apt.apt_pkg.config.clear('APT::Update::Post-Invoke-Success')
+
+        self.progress = apt.progress.text.AcquireProgress()
+        if not os.isatty(1):
+            # Make output more suitable for logging.
+            self.progress.pulse = lambda owner: True
+            self.progress._width = 0
+
+    def _setup_apt_cache(self, rootdir):
+        if self._use_geoip or self._sources_list:
+            release = platform.linux_distribution()[2]
+            sources_list = _format_sources_list(
+                self._sources_list, deb_arch=self._deb_arch,
+                use_geoip=self._use_geoip, release=release)
+        else:
+            sources_list = _get_local_sources_list()
+
+        sources_list_digest = hashlib.sha384(
+            sources_list.encode(sys.getfilesystemencoding())).hexdigest()
+
+        cache_dir = os.path.join(self._cache_dir, sources_list_digest)
+        apt_cache_dir = os.path.join(cache_dir, 'apt')
+        package_cache_dir = os.path.join(cache_dir, 'packages')
+
+        sources_list_file = os.path.join(
+            apt_cache_dir, 'etc', 'apt', 'sources.list')
+
+        os.makedirs(os.path.dirname(sources_list_file), exist_ok=True)
+        with open(sources_list_file, 'w') as f:
+            f.write(sources_list)
+
+        apt_cache = apt.Cache(rootdir=apt_cache_dir, memonly=True)
+        apt_cache.update(fetch_progress=self.progress,
+                         sources_list=sources_list_file)
+
+        copy_tree(apt_cache_dir, rootdir, update=True)
+
+        self.apt_cache = apt.Cache(rootdir=rootdir, memonly=True)
+        self.apt_cache.open()
+
+        return package_cache_dir
```

Comment:
El 10/08/16 a las 13:23, Leo Arias escribió:
> In snapcraft/internal/repo.py
> <https://github.com/snapcore/snapcraft/pull/677#discussion_r74278577>:
> 
>> +            apt_cache_dir, 'etc', 'apt', 'sources.list')
>> +
>> +        os.makedirs(os.path.dirname(sources_list_file), exist_ok=True)
>> +        with open(sources_list_file, 'w') as f:
>> +            f.write(sources_list)
>> +
>> +        apt_cache = apt.Cache(rootdir=apt_cache_dir, memonly=True)
>> +        apt_cache.update(fetch_progress=self.progress,
>> +                         sources_list=sources_list_file)
>> +
>> +        copy_tree(apt_cache_dir, rootdir, update=True)
>> +
>> +        self.apt_cache = apt.Cache(rootdir=rootdir, memonly=True)
>> +        self.apt_cache.open()
>> +
>> +        return package_cache_dir
> 
> what about returning a tuple (package_cache_dir, apt_cache) ? :)

Then apt_cache needs to be passed around as well.